Sources/MongoSwift/BSON/BSONValue.swift

```diff
@@ -197,50 +117,33 @@ extension Array: BSONValue {
             }
 
             let arrDoc = Document(stealing: arrayData)
-
-            guard let arr = arrDoc.values as? Array else {
-                fatalError("Failed to cast values for document \(arrDoc) to array")
-            }
-
-            return arr
-        }
+            return arrDoc.values
+        })
     }
 
-    public func encode(to storage: DocumentStorage, forKey key: String) throws {
+    internal func encode(to storage: DocumentStorage, forKey key: String) throws {
         var arr = Document()
         for (i, v) in self.enumerated() {
-            guard let val = v as? BSONValue else {
-                throw UserError.logicError(
-                    message: "Cannot encode a non-BSONValue array element: \(String(describing: v)) "
-                        + "with type: \(type(of: v)) "
-                        + "at index: \(i)"
-                )
-            }
-            try arr.setValue(for: String(i), to: val)
+            try arr.setValue(for: String(i), to: v)
         }
 
         guard bson_append_array(storage._bson, key, Int32(key.utf8.count), arr._bson) else {
             throw bsonTooLargeError(value: self, forKey: key)
         }
     }
-
-    public func bsonEquals(_ other: BSONValue?) -> Bool {
-        guard let otherArr = other as? [BSONValue], let selfArr = self as? [BSONValue] else {
-            return false
-        }
-        return self.count == otherArr.count && zip(selfArr, otherArr).allSatisfy { lhs, rhs in lhs.bsonEquals(rhs) }
-    }
 }
 
 /// A struct to represent the BSON null type.
-public struct BSONNull: BSONValue, Codable, Equatable {
-    public var bsonType: BSONType { return .null }
+internal struct BSONNull: BSONValue, Codable, Equatable {
```

Comment:
users have no need of this struct anymore unless they wan't to specifically decode `BSONNull` as a field in their structs, but that doesn't seem like a common or useful case. So now all the `BSONValue`s that aren't included in `BSON` are internal (`BSONNull`, `BSONUndefined`, `MinKey`, `MaxKey`).